src/Covenant/JSON.hs

```diff
@@ -86,44 +268,94 @@ import Data.Text qualified as T
 import Data.Vector (Vector)
 import Data.Vector qualified as Vector
 import Data.Vector.NonEmpty qualified as NEV
-import Data.Word (Word32)
+import Data.Void (Void, absurd)
+import Data.Wedge (Wedge (Here, Nowhere, There))
 import GHC.TypeLits (KnownSymbol, Symbol)
-import Optics.Core (preview, review)
+import Optics.Core (preview, review, set, view)
 import Text.Hex qualified as Hex
 
--- | A concrete monadic stack, providing the minimum amount of functionality
--- needed to build an ASG using the combinators given in this module, plus
--- some functionality needed to handle parsing an ASG from JSON.
--- @since 1.3.0
-newtype ASGParser (a :: Type)
-  = ASGParser (ReaderT ASGEnv (ExceptT CovenantTypeError (HashConsT Id ASGNode Parser)) a)
-  deriving
-    ( -- | @since 1.3.0
-      Functor,
-      -- | @since 1.3.0
-      Applicative,
-      -- | @since 1.3.0
-      Monad,
-      -- | @since 1.3.0
-      MonadReader ASGEnv,
-      -- | @since 1.3.0
-      MonadError CovenantTypeError,
-      -- | @since 1.3.0
-      MonadHashCons Id ASGNode
-    )
-    via ReaderT ASGEnv (ExceptT CovenantTypeError (HashConsT Id ASGNode Parser))
+compileAndSerialize ::
+  forall (a :: Type).
+  FilePath ->
+  [DataDeclaration AbstractTy] ->
+  ASGBuilder a ->
+  Version ->
+  IO ()
+compileAndSerialize path decls asgBuilder version = do
+  case mkDatatypeInfos decls of
+    Left err' -> throwIO . userError $ err'
+    Right infos -> case runASGBuilder infos asgBuilder of
+      Left err' -> throwIO . userError . show $ err'
+      Right (ASG (_, asg)) -> do
+        let cu = CompilationUnit (Vector.fromList decls) asg version
+        writeJSONWith path cu encodeCompilationUnit
+
+deserializeAndValidate ::
+  FilePath ->
+  IO ASG
+deserializeAndValidate path = do
+  rawCU <- readJSON @CompilationUnit path
+  case validateCompilationUnit rawCU of
+    Left err' -> throwIO . userError $ "Failed to validate ASG. Reason:\n" <> show err'
+    Right asg -> pure asg
 
 data Version = Version {_major :: Int, _minor :: Int}
   deriving stock (Show, Eq, Ord)
 
 data CompilationUnit
   = CompilationUnit
-  { _datatypes :: Map TyName (DataDeclaration AbstractTy),
+  { _datatypes :: Vector (DataDeclaration AbstractTy),
     _asg :: Map Id ASGNode,
     _version :: Version
   }
   deriving stock (Show, Eq)
```

Comment:
I question having `CompilationUnit` be completely open like this. I understand that `compileAndSerialize` technically protects us from the worst of it, but this interface feels kludgey.